File: src/agent/history/storage.py
"""Conversation history persistence as JSONL files (one JSON object per line)."""

from __future__ import annotations

import json
import uuid
from datetime import datetime
from pathlib import Path
from typing import Any

from agent.core.message import Message


class HistoryStorage:
    """Saves and loads conversation histories as JSONL.

    Format:
      Line 1: header  {"type": "header", "session_id": "...", "timestamp": "...", "metadata": {...}}
      Line 2+: messages {"type": "message", "role": "...", "content": ..., "token_count": N}

    Appending new messages is O(new_messages) — no need to rewrite the file.
    """

    def __init__(self, history_dir: str, base_dir: Path) -> None:
        self._dir = base_dir / history_dir
        self._dir.mkdir(parents=True, exist_ok=True)

    def new_session_id(self) -> str:
        return datetime.now().strftime("%Y%m%d_%H%M%S") + "_" + uuid.uuid4().hex[:8]

    def save(self, session_id: str, messages: list[Message], metadata: dict[str, Any] | None = None) -> Path:
        """Append new messages to a JSONL session file.

        Only messages not yet written are appended, making repeated saves efficient.
        """
        path = self._dir / f"{session_id}.jsonl"
        existing_count = 0

        if path.exists():
            # Count existing message lines (skip header)
            with path.open("r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        obj = json.loads(line)
                        if obj.get("type") == "message":
                            existing_count += 1
                    except json.JSONDecodeError:
                        continue
        else:
            # Write header as first line
            header = {
                "type": "header",
                "session_id": session_id,
                "timestamp": datetime.now().isoformat(),
                "metadata": metadata or {},
            }
            with path.open("w", encoding="utf-8") as f:
                f.write(json.dumps(header, ensure_ascii=False) + "\n")

        # Append only new messages
        new_messages = messages[existing_count:]
        if new_messages:
            with path.open("a", encoding="utf-8") as f:
                for msg in new_messages:
                    record = {
                        "type": "message",
                        "role": msg.role,
                        "content": msg.content,
                        "token_count": msg.token_count,
                    }
                    f.write(json.dumps(record, ensure_ascii=False) + "\n")

        return path

    def rewrite(self, session_id: str, messages: list[Message], metadata: dict[str, Any] | None = None) -> Path:
        """Rewrite a session file from scratch (e.g. after compaction).

        Unlike save(), this replaces the entire file instead of appending.
        """
        path = self._dir / f"{session_id}.jsonl"
        header = {
            "type": "header",
            "session_id": session_id,
            "timestamp": datetime.now().isoformat(),
            "metadata": metadata or {},
        }
        with path.open("w", encoding="utf-8") as f:
            f.write(json.dumps(header, ensure_ascii=False) + "\n")
            for msg in messages:
                record = {
                    "type": "message",
                    "role": msg.role,
                    "content": msg.content,
                    "token_count": msg.token_count,
                }
                f.write(json.dumps(record, ensure_ascii=False) + "\n")
        return path

    def load(self, session_id: str) -> list[Message] | None:
        """Load a conversation from a JSONL file."""
        path = self._dir / f"{session_id}.jsonl"
        if not path.exists():
            # Fallback: try legacy .json format
            return self._load_legacy_json(session_id)

        messages = []
        with path.open("r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if obj.get("type") != "message":
                    continue
                msg = Message(
                    role=obj["role"],
                    content=obj["content"],
                    token_count=obj.get("token_count", 0),
                )
                messages.append(msg)
        return messages if messages else None

    def _load_legacy_json(self, session_id: str) -> list[Message] | None:
        """Load from old .json format for backwards compatibility."""
        path = self._dir / f"{session_id}.json"
        if not path.exists():
            return None
        data = json.loads(path.read_text())
        messages = []
        for msg_data in data.get("messages", []):
            msg = Message(
                role=msg_data["role"],
                content=msg_data["content"],
                token_count=msg_data.get("token_count", 0),
            )
            messages.append(msg)
        return messages

    def clear_session(self, session_id: str) -> None:
        """Reset a session file to header-only (no messages)."""
        path = self._dir / f"{session_id}.jsonl"
        if not path.exists():
            return
        # Read header, rewrite file with header only
        header = None
        with path.open("r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                    if obj.get("type") == "header":
                        header = obj
                        break
                except json.JSONDecodeError:
                    continue
        if header:
            with path.open("w", encoding="utf-8") as f:
                f.write(json.dumps(header, ensure_ascii=False) + "\n")

    def find_session(self, query: str) -> str | None:
        """Find a session ID by exact or prefix match. Returns full ID or None."""
        # Exact match (prefer .jsonl, fallback to .json)
        if (self._dir / f"{query}.jsonl").exists():
            return query
        if (self._dir / f"{query}.json").exists():
            return query
        # Prefix match
        matches = sorted(
            list(self._dir.glob(f"{query}*.jsonl")) + list(self._dir.glob(f"{query}*.json")),
            key=lambda f: f.stat().st_mtime,
            reverse=True,
        )
        if matches:
            return matches[0].stem
        return None

    def get_latest_session_id(self) -> str | None:
        """Return the most recent session ID, or None if no sessions exist."""
        files = sorted(
            list(self._dir.glob("*.jsonl")) + list(self._dir.glob("*.json")),
            key=lambda f: f.stat().st_mtime,
            reverse=True,
        )
        return files[0].stem if files else None

    def list_sessions(self, limit: int = 20) -> list[dict[str, Any]]:
        """List recent sessions (newest first)."""
        files = sorted(
            list(self._dir.glob("*.jsonl")) + list(self._dir.glob("*.json")),
            key=lambda f: f.stat().st_mtime,
            reverse=True,
        )
        sessions = []
        for f in files[:limit]:
            try:
                if f.suffix == ".jsonl":
                    sessions.append(self._session_info_jsonl(f))
                else:
                    sessions.append(self._session_info_json(f))
            except Exception:
                continue
        return sessions

    def _session_info_jsonl(self, path: Path) -> dict[str, Any]:
        """Extract session info from a JSONL file (reads only what's needed)."""
        session_id = path.stem
        timestamp = ""
        message_count = 0
        with path.open("r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if obj.get("type") == "header":
                    session_id = obj.get("session_id", session_id)
                    timestamp = obj.get("timestamp", "")
                elif obj.get("type") == "message":
                    message_count += 1
        return {
            "session_id": session_id,
            "timestamp": timestamp,
            "message_count": message_count,
        }

    def _session_info_json(self, path: Path) -> dict[str, Any]:
        """Extract session info from a legacy JSON file."""
        data = json.loads(path.read_text())
        return {
            "session_id": data.get("session_id", path.stem),
            "timestamp": data.get("timestamp", ""),
            "message_count": len(data.get("messages", [])),
        }

File: src/history/storage.ts
/** Conversation history persistence as JSONL files (one JSON object per line). */

import { randomBytes } from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

import { Message, type ContentBlock } from '../core/message.js';

interface SessionInfo {
  session_id: string;
  timestamp: string;
  message_count: number;
}

function timestamp(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, '0');
  return (
    `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}_` +
    `${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`
  );
}

/**
 * Saves and loads conversation histories as JSONL.
 *
 *   Line 1:  header   {"type":"header","session_id":...,"timestamp":...,"metadata":...}
 *   Line 2+: messages {"type":"message","role":...,"content":...,"token_count":N}
 */
export class HistoryStorage {
  private dir: string;

  constructor(historyDir: string, baseDir: string) {
    this.dir = path.join(baseDir, historyDir);
    fs.mkdirSync(this.dir, { recursive: true });
  }

  newSessionId(): string {
    return `${timestamp()}_${randomBytes(4).toString('hex')}`;
  }

  private jsonlPath(sessionId: string): string {
    return path.join(this.dir, `${sessionId}.jsonl`);
  }

  private jsonPath(sessionId: string): string {
    return path.join(this.dir, `${sessionId}.json`);
  }

  /** Append new messages to a JSONL session file. */
  save(
    sessionId: string,
    messages: Message[],
    metadata: Record<string, any> | null = null,
  ): string {
    const filePath = this.jsonlPath(sessionId);
    let existingCount = 0;

    if (fs.existsSync(filePath)) {
      const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const obj = JSON.parse(trimmed);
          if (obj.type === 'message') existingCount += 1;
        } catch {
          /* ignore */
        }
      }
    } else {
      const header = {
        type: 'header',
        session_id: sessionId,
        timestamp: new Date().toISOString(),
        metadata: metadata ?? {},
      };
      fs.writeFileSync(filePath, JSON.stringify(header) + '\n');
    }

    const newMessages = messages.slice(existingCount);
    if (newMessages.length > 0) {
      const lines = newMessages.map((msg) =>
        JSON.stringify({
          type: 'message',
          role: msg.role,
          content: msg.content,
          token_count: msg.tokenCount,
        }),
      );
      fs.appendFileSync(filePath, lines.join('\n') + '\n');
    }

    return filePath;
  }

  /** Rewrite a session file from scratch (e.g. after compaction). */
  rewrite(
    sessionId: string,
    messages: Message[],
    metadata: Record<string, any> | null = null,
  ): string {
    const filePath = this.jsonlPath(sessionId);
    const header = {
      type: 'header',
      session_id: sessionId,
      timestamp: new Date().toISOString(),
      metadata: metadata ?? {},
    };
    const lines = [JSON.stringify(header)];
    for (const msg of messages) {
      lines.push(
        JSON.stringify({
          type: 'message',
          role: msg.role,
          content: msg.content,
          token_count: msg.tokenCount,
        }),
      );
    }
    fs.writeFileSync(filePath, lines.join('\n') + '\n');
    return filePath;
  }

  load(sessionId: string): Message[] | null {
    const filePath = this.jsonlPath(sessionId);
    if (!fs.existsSync(filePath)) {
      return this.loadLegacyJson(sessionId);
    }
    const messages: Message[] = [];
    const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      let obj: any;
      try {
        obj = JSON.parse(trimmed);
      } catch {
        continue;
      }
      if (obj.type !== 'message') continue;
      messages.push(new Message(obj.role, obj.content as ContentBlock[], obj.token_count ?? 0));
    }
    return messages.length > 0 ? messages : null;
  }

  private loadLegacyJson(sessionId: string): Message[] | null {
    const filePath = this.jsonPath(sessionId);
    if (!fs.existsSync(filePath)) return null;
    const data = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
    const messages: Message[] = [];
    for (const m of data.messages ?? []) {
      messages.push(new Message(m.role, m.content as ContentBlock[], m.token_count ?? 0));
    }
    return messages;
  }

  /** Reset a session file to header-only (no messages). */
  clearSession(sessionId: string): void {
    const filePath = this.jsonlPath(sessionId);
    if (!fs.existsSync(filePath)) return;
    let header: any = null;
    const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      try {
        const obj = JSON.parse(trimmed);
        if (obj.type === 'header') {
          header = obj;
          break;
        }
      } catch {
        /* ignore */
      }
    }
    if (header) {
      fs.writeFileSync(filePath, JSON.stringify(header) + '\n');
    }
  }

  private listFiles(): Array<{ file: string; stem: string; mtime: number; ext: string }> {
    let entries: string[];
    try {
      entries = fs.readdirSync(this.dir);
    } catch {
      return [];
    }
    const files: Array<{ file: string; stem: string; mtime: number; ext: string }> = [];
    for (const name of entries) {
      const ext = path.extname(name);
      if (ext !== '.jsonl' && ext !== '.json') continue;
      const full = path.join(this.dir, name);
      let mtime = 0;
      try {
        mtime = fs.statSync(full).mtimeMs;
      } catch {
        continue;
      }
      files.push({ file: full, stem: path.basename(name, ext), mtime, ext });
    }
    files.sort((a, b) => b.mtime - a.mtime);
    return files;
  }

  /** Find a session ID by exact or prefix match. Returns full ID or null. */
  findSession(query: string): string | null {
    if (fs.existsSync(this.jsonlPath(query))) return query;
    if (fs.existsSync(this.jsonPath(query))) return query;
    const matches = this.listFiles().filter((f) => f.stem.startsWith(query));
    return matches.length > 0 ? matches[0]!.stem : null;
  }

  getLatestSessionId(): string | null {
    const files = this.listFiles();
    return files.length > 0 ? files[0]!.stem : null;
  }

  /** List recent sessions (newest first). */
  listSessions(limit = 20): SessionInfo[] {
    const files = this.listFiles().slice(0, limit);
    const sessions: SessionInfo[] = [];
    for (const f of files) {
      try {
        sessions.push(
          f.ext === '.jsonl' ? this.sessionInfoJsonl(f.file) : this.sessionInfoJson(f.file),
        );
      } catch {
        continue;
      }
    }
    return sessions;
  }

  private sessionInfoJsonl(filePath: string): SessionInfo {
    let sessionId = path.basename(filePath, '.jsonl');
    let ts = '';
    let messageCount = 0;
    const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      let obj: any;
      try {
        obj = JSON.parse(trimmed);
      } catch {
        continue;
      }
      if (obj.type === 'header') {
        sessionId = obj.session_id ?? sessionId;
        ts = obj.timestamp ?? '';
      } else if (obj.type === 'message') {
        messageCount += 1;
      }
    }
    return { session_id: sessionId, timestamp: ts, message_count: messageCount };
  }

  private sessionInfoJson(filePath: string): SessionInfo {
    const data = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
    return {
      session_id: data.session_id ?? path.basename(filePath, '.json'),
      timestamp: data.timestamp ?? '',
      message_count: (data.messages ?? []).length,
    };
  }
}
